src/fancy-tools/freehandTool.js

```diff
@@ -0,0 +1,1143 @@
+      invalidHandlePlacement = false;
+    } else {
+      invalidHandlePlacement = freeHandIntersect.newHandle(mousePoint, dataHandles);
+    }
+
+    return invalidHandlePlacement;
+  }
+
+  /**
+  * Returns true if the proposed location of a new handle is invalid (in pencilMode).
+  *
+  * @private
+  * @param {Object} data - data object associated with the tool.
+  * @return {Boolean}
+  */
+  _checkHandlesPencilMode (data) {
+    const config = this.configuration;
+    const mousePoint = config.mouseLocation.handles.start;
+    const dataHandles = data.handles;
+    let invalidHandlePlacement = freeHandIntersect.newHandle(mousePoint, dataHandles);
+
+    if (invalidHandlePlacement === false) {
+      invalidHandlePlacement = this._invalidHandlePencilMode(data, mousePoint);
+    }
+
+    return invalidHandlePlacement;
+  }
+
+  /**
+  * Returns true if the mouse position is far enough from previous points (in pencilMode).
+  *
+  * @private
+  * @param {Object} data - data object associated with the tool.
+  * @param {Object} mousePoint - the position of the mouse cursor.
+  * @return {Boolean}
+  */
+  _invalidHandlePencilMode (data, mousePoint) {
+    const config = this.configuration;
+    const dataHandles = data.handles;
+
+    if (external.cornerstoneMath.point.distance(dataHandles[0], mousePoint) < config.spacing) {
+      data.canComplete = true;
+
+      return false;
+    }
+
+    data.canComplete = false;
+
+    // Compare with all other handles appart from the last one
+    for (let i = 1; i < dataHandles.length - 1; i++) {
+      if (external.cornerstoneMath.point.distance(dataHandles[i], mousePoint) < config.spacing) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+  * Activates a particular tool when the mouseCursor is near one if it's handles
+  * or it's textBox.
+  *
+  * @private
+  * @static
+  * @param {Object} eventData - The data assoicated with the event.
+  * @param {Object} toolData - The data associated with the tool.
+  */
+  static _mouseHover (eventData, toolData) {
+    // Check if user is mousing over the textBox
+    let imageNeedsUpdate = false;
+
+    for (let i = 0; i < toolData.data.length; i++) {
+      // Get the cursor position in canvas coordinates
+      const data = toolData.data[i];
+      const coords = eventData.currentPoints.canvas;
+
+      if (data.textBox === true) {
+        if (pointInsideBoundingBox(data.handles.textBox, coords)) {
+          data.active = !data.active;
+          data.highlight = !data.highlight;
+          imageNeedsUpdate = true;
+        }
+      }
+    }
+
+    return imageNeedsUpdate;
+  }
+
+  /**
+  * Adds commas as thousand seperators to a Number to increase readability.
+  *
+  * @private
+  * @static
+  * @param {Number|String} number - A Number or String literal representing a number.
+  * @return {String} - A string literal representaton of the number with commas seperating the thousands.
+  */
+  static _numberWithCommas (number) {
```

Comment:
`rectangleRoi` and `ellipticalRoi` have this same method. We should move it to shared/utils and import.

Reply:
I see the two relevant PRs are merged, I'll wire this up to all three.

Reply:
Done an pushed.